### ISSUE TITLE
Massive Bottleneck in window_masking due to CPU-GPU Synchronization and NumPy calls

### ISSUE THREAD

**Issue by WenMing8888** [Zian-Xu/Swin-MAE#24]
While performing large-scale pre-training (5.2M images) using Swin-MAE on high-end hardware (RTX 5090), I identified a severe performance bottleneck in the window_masking function.
The implementation causes the GPU utilization to fluctuate wildly (dropping to ~30% or lower) because of frequent implicit CPU-GPU synchronizations and slow Python-native loops during the forward pass.
Problem Breakdown
The following lines in the current implementation are the primary cause of the performance degradation:
Implicit Synchronization: The call to .cpu().numpy() inside the forward loop forces the GPU to halt and wait for data transfer over the PCIe bus.
CPU Logic Bottleneck: Using np.setdiff1d and Python for loops to calculate mask indices is extremely slow compared to the GPU's computation speed. On an RTX 5090, the GPU finishes the batch calculation in milliseconds, while this CPU logic takes ~2 seconds.
Redundant Memory Allocation: Frequent creation of large NumPy arrays and re-conversion to Torch tensors adds significant overhead.

**Comment by Zian-Xu** [comment]
Thank you so much for your thorough analysis and for taking the time to identify potential performance bottlenecks in the window_masking function. We really appreciate your detailed breakdown.

I would like to note that the Swin-MAE code was written about four years ago, at a time when GPUs were not nearly as powerful as the current high-end devices like the RTX 5090. At that time, such performance bottlenecks were not noticeable. Additionally, the original implementation of Swin-MAE was not primarily optimized for runtime speed; the main focus was on correctness and demonstrating the Masked Autoencoder approach.

Nevertheless, your observations are very valuable, and it’s good to be aware of potential slowdowns in large-scale pre-training scenarios. Thank you again for pointing this out!

**Comment by BorisMorozov** [comment]
Hey! I've been looking for GPU stalling issues in real codebases to test Ingero in real-life scenarios. This issue seemed like a perfect match, so I decided to dig in.

I managed to fix it by dropping the NumPy ops and keeping everything in pure PyTorch:

```python
# Before: B × 2 implicit cudaStreamSync calls per forward pass
for i in range(B):
    index_mask[i] = np.setdiff1d(index_all, index_keep.cpu().numpy()[i])
for i in range(B):
    x_masked[i, index_mask.cpu().numpy()[i, :], :] = mask_token

# After: Zero CPU transfers!
index_mask = ids_shuffle[:, -mask_len:]
x_masked.scatter_(1, index_mask.unsqueeze(-1).expand(-1, -1, C), mask_token)
```

When I tested this fix, the throughput increased
~6.4x speedup on a T4 (AWS EC2)
~13x speedup on an RTX 5080

---
(Shameless plug / full disclosure): Since I mentioned it above, I'm actually one of the co-creators of [Ingero](https://github.com/ingero-io/ingero). If you're ever interested in effortlessly diagnosing GPU stalling issues like this one, give it a try! You can use it the old-school way by manually inspecting the detected causal chains, or you can hook it up to an LLM agent via Ingero MCP.
